### PR TITLE
Extend TCanvasImp interface

### PR DESCRIPTION
**Pull request by linev**
- let create TVirtualPadPainter via the TCanvasImp.
- handling of TCanvas::Update() can be perfromed in TCanvasImp

### CODE REVIEW

**Comment by bellenot** [comment]
I will merge once the tests are over...

**Comment by linev** [comment]
Usual question - should I reformat original TCanvas code, which I kept as is.

**Comment by pcanal** [comment]
yes in this case, for 'just' the line changed (i.e. the travis recommendation).

**Comment by linev** [comment]
Done

**Comment by linev** [comment]
> yes in this case, for 'just' the line changed (i.e. the travis recommendation).

Philippe, your command does not work for me. I tried:

     git-clang-format --commit 31fb30ff38 --diff --binary `which clang-format`

**Comment by pcanal** [comment]
@linev How does it fail?
@vgvassilev Can you help me?

**Comment by linev** [comment]
> How does it fail?

It says that there is nothing changed.

**Comment by pcanal** [comment]
Note: the two failing test (math and cling) are unrelated to this PR.

**Comment by linev** [comment]
> How does it fail?

It says that there is nothing changed.

**Comment by pcanal** [comment]
@vgvassilev any ideas?

**Comment by vgvassilev** [comment]
'projectroot.math.mathcore.test.mathcore_testLogLExecPolicy' should be fixed by #627.
'projectroot.roottest.python.stl.roottest_python_stl_stl' should be fixed by #614.

The other failure I don't know.

File: graf2d/gpad/src/TCanvas.cxx [hunk 1]
```diff
@@ -2299,9 +2302,10 @@ void TCanvas::CreatePainter()
 {
    //Even for batch mode painter is still required, just to delegate
    //some calls to batch "virtual X".
-   if (!UseGL() || fBatch)
-      fPainter = new TPadPainter;//Do not need plugin manager for this!
-   else {
+   if (!UseGL() || fBatch) {
+      fPainter = !fCanvasImp ? 0 : fCanvasImp->CreatePadPainter();
```
Comment (Axel-Naumann) [inline graf2d/gpad/src/TCanvas.cxx]: Is
```
if (fCanvasImp)
  fPainter = fCanvasImp->CreatePadPainter();
```
maybe a more readable version of this code?

Comment (linev) [inline graf2d/gpad/src/TCanvas.cxx]: > maybe a more readable version of this code?

Done

**Pull request merged**